sensor/common/compliance/nodescan_handler_test.go

```diff
@@ -51,85 +48,138 @@ func fakeNodeScanV2(nodeName string) *storage.NodeScanV2 {
 
 type NodeScanHandlerTestSuite struct {
 	suite.Suite
-	cancel    context.CancelFunc
-	ctx       context.Context
-	toCentral chan *central.MsgFromSensor
+	cancel context.CancelFunc
+	ctx    context.Context
 }
 
 func (s *NodeScanHandlerTestSuite) SetupTest() {
-	s.ctx, s.cancel = context.WithCancel(context.Background())
-	s.toCentral = make(chan *central.MsgFromSensor)
+
+}
+
+func assertNoGoroutineLeaks(t *testing.T) {
+	goleak.VerifyNone(t,
+		// Ignore a known leak: https://github.com/DataDog/dd-trace-go/issues/1469
+		goleak.IgnoreTopFunction("github.com/golang/glog.(*loggingT).flushDaemon"),
+	)
 }
 
 func (s *NodeScanHandlerTestSuite) TearDownTest() {
 	defer assertNoGoroutineLeaks(s.T())
 	s.cancel()
 }
 
-func (s *NodeScanHandlerTestSuite) startForwardingToCentral(c <-chan *central.MsgFromSensor) {
+// TestStopHandler goal is to stop handler while there are still some messages to process
+// in the channel passed into NewNodeScanHandler.
+// We expect that premature stop of the handler produces no race condition or goroutine leak.
+// Exec with: go test -race -count=1 -v -run ^TestNodeScanHandler$ ./sensor/common/compliance
+func (s *NodeScanHandlerTestSuite) TestStopHandler() {
+	nodeScans := make(chan *storage.NodeScanV2)
+	h := NewNodeScanHandler(nodeScans)
+	s.consumeToCentral(h)
+	// This is a producer that stops the handler after producing the first message and then sends many (29) more messages
+	// The intent here it to test whether the handler can shutdown with no leaks when the context is canceled despite of
+	// multiple messages waiting in the channel to be processed
 	go func() {
-		defer close(s.toCentral)
-		for {
+		defer close(nodeScans)
```

Comment:
> Closing the channel in this goroutine is interesting and maybe asks for a dedicated test.

I am following a pattern here where the writer is responsible for closing the channel that it writes to when its writing-work is done.

> Depending on what select statement in run() sees first - closed channel, or signalled stopC, the result returned from h.Stopped() will have or not have an error.

It will have error in both cases, just the error will be different - either the error passed as an argument to `Stop()` (which can be nil), or the predefined `errInputChanClosed`. 

> For this test I suggest to move defer close(nodeScans) outside of this goroutine. That way, it will be called after the node scan handler reports that it's stopped.

Yes, that should allow us to make an `ErrorIs` assertion without the or-condition, but there is a risk of writing to the channel after closing it. I will check whether the risk can be eliminated in that particular test case.

Reply:
I checked that and not closing the channel in the goroutine triggers a data race between the test routine and the anonymous routine. I would need to add a WaitGroup to prevent this.

This is how this case would look like. It allows us to assert `s.NoError(h.Stopped().Wait())` but it clutters the code a bit more.

```go
func (s *NodeScanHandlerTestSuite) TestHandlerRegularRoutine2() {
	wg := &sync.WaitGroup{}
	wg.Add(1)
	ch := s.generateTestInputNoClose(wg)
	defer func(wg *sync.WaitGroup) {
		wg.Wait()
		close(ch)
	}(wg)
	h := NewNodeScanHandler(ch)
	s.NoError(h.Start())
	s.consumeToCentral(h)
	h.Stop(nil)
	s.NoError(h.Stopped().Wait())
}
```

Will not commit this for now and wait for your reply.